rfd/001-tcp-load-balancer-design-doc.md

```diff
@@ -0,0 +1,119 @@
+
+```
+# Generate serverca.cert and servercakey.pem to allow signing of server keys.
+$ openssl genrsa -out servercakey.pem
+$ openssl req -new -x509 -key servercakey.pem -out serverca.cert
+
+# Generate public (server.key) and private (server.crt) keys for the the server.
+$ openssl genrsa -out server.key
+$ openssl req -new -key server.key -out server_reqout.txt
+$ openssl x509 -req -in server_reqout.txt -days 3650 -sha256 -CAcreateserial -CA serverca.cert -CAkey servercakey.pem -out server.crt
+
+# Generate clentca.cert and clientcakey.pem to allow signing of client keys.
+$ openssl genrsa -out clientcakey.pem
+$ openssl req -new -x509 -key clientcakey.pem -out clentca.cert
+
+Generate private (client.crt) and public (client.key) keys for the client:
+$ openssl genrsa -out client.key
+$ openssl req -new -key client.key -out client_reqout.txt
+$ openssl x509 -req -in client_reqout.txt -days 3650 -sha256 -CAcreateserial -CA clentca.cert -CAkey clientcakey.pem -out client.crt		
+```
+
+For this project, a test client will be created and provided with a valid client cert for testing purposes. These initial dummy certificates and keys will be included (as plaintext) within this repository for the the benefit of reviewers, but should stored as secure secrets and managed outside of this repo in a future version.
+
+Additionally, the requirement of over-the-wire encryption is not mentioned, so even though the authentication requires certs and keys, the data sent after authentication may or may not be encrypted depending on what is most convenient at the point of implementation.
+
+### Rate Limiter
+This load balancer will implement a per-client connection rate limiter that tracks the number of client connections, and limits to n (configurable) active connections per client. It could track client identity based on attributes like IP address, and device identifier, but since this load balancer requires authentication (mTLS), we will identify unique clients based on a V5 UUID generated from the client's TLS certificate. 
+
+```
+clientID := uuid.NewV5(loadBalancerID, client.TLSCertificate().String())
+```
+
+Client connection counts can be stored in a `sync.Map` structure for quick modification, and safety across concurrent goroutines. Regarding this map choice, the [`sync` package documentation](https://pkg.go.dev/sync#Map) states the following: 
+> "The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex."
+
+Our use-case follows the #2 reason for leveraging `sync.Map` over a standard map with a Mutex; our disjoint sets of keys will be the unique client ids.
+
+
+### Upstream Authorization Management
```

Comment:
Could you share a simple example of an authorization scheme?

Reply:
Yes, I'll get that example added to the doc. Here's what I'm thinking for a re-write of this section (to include a better example):

### Upstream Authorization Management

The load balancer will employ a simple authorization scheme which will deny access to specific upstreams for certain clients. This scheme will be statically defined in code, using something like the following `map` of `UnauthorizedClientHostRule`s to ensure an authorized host is selected. To simply demonstrate this capability, we store a map of `clientID:hostID` as an `UnauthorizedClientHostRule`. The only example rule which will be present for this project will be that client1 can never talk to host2. 

```

// UnauthorizedClientHostRule is a composite key made up of ClientID and HostID, and represents that the client is not authorized to connect to the host.
type UnauthorizedClientHostRule string

// FormatUnauthorizedClientHostRule builds a new unauthorizedClientHostRule from the given client and host IDs.
func BuildUnauthorizedClientHostRule(clientID uuid.UUID, hostID uuid.UUID) UnauthorizedClientHostRule {
	return UnauthorizedClientHostRule(fmt.Sprintf("%s:%s", clientID, hostID))
}

// unauthorized serves as a key-based lookup for unauthorized client-> host relationships.
// It will be checked every time a host is being selected. 
// If upstream hosts are to be registered dynamically, access to this map should be locked behind a mutex.
// For simplicity of this project, we'll only ever add rules during app startup, so the map is safe for concurrent reads.
var unauthorized = map[UnauthorizedClientHostRule]struct{}{}
```

Once the load balancer has determined that the client is authenticated and within client connection limit (and prior to selecting the least connections host) it will filter out unauthorized hosts for that client. See the following section for how the unauthorized map may be used for connection filtering.

...
(Then the following section will have the code example update checking the unauthorized map which looks like this)
```
func (l *LoadBalancer) LeastConnections(clientID uuid.UUID) *upstream.TcpHost {
	if l == nil || len(l.hosts) == 0 {
		return nil
	}

	var host *upstream.TcpHost

	for i, h := range l.hosts {
         _, unauthzd := unauthorized[BuildUnauthorizedClientHostRule(clientID, h.ID)]
		if !unauthzd && (host == nil || h.ConnectionCount() < host.ConnectionCount()) {
			host = h
		}
	}

	return host
}
```

Reply:
Does that cover the example you're looking for?

Reply:
👍  Looks reasonable. Thanks.